File: src/python/graphql/graphene/schema.py
import uuid
from datetime import datetime, timezone

import graphene

from db import check_db, get_user, get_complex_orders
from cache import check_cache, cache_get, cache_set


class Health(graphene.ObjectType):
    status = graphene.NonNull(graphene.String)
    version = graphene.NonNull(graphene.String)
    timestamp = graphene.NonNull(graphene.String)
    database = graphene.NonNull(graphene.String)
    cache = graphene.NonNull(graphene.String)


class JsonItem(graphene.ObjectType):
    id = graphene.NonNull(graphene.Int)
    uuid = graphene.NonNull(graphene.String)
    name = graphene.NonNull(graphene.String)
    email = graphene.NonNull(graphene.String)
    created_at = graphene.NonNull(graphene.String, name="createdAt")
    is_active = graphene.NonNull(graphene.Boolean, name="isActive")


class JsonItemsResult(graphene.ObjectType):
    items = graphene.NonNull(graphene.List(graphene.NonNull(JsonItem)))
    count = graphene.NonNull(graphene.Int)
    timestamp = graphene.NonNull(graphene.String)


class User(graphene.ObjectType):
    id = graphene.NonNull(graphene.Int)
    email = graphene.NonNull(graphene.String)
    first_name = graphene.NonNull(graphene.String, name="firstName")
    last_name = graphene.NonNull(graphene.String, name="lastName")
    age = graphene.NonNull(graphene.Int)
    created_at = graphene.NonNull(graphene.String, name="createdAt")


class UserOrderStats(graphene.ObjectType):
    user_id = graphene.NonNull(graphene.Int, name="userId")
    user_name = graphene.NonNull(graphene.String, name="userName")
    total_orders = graphene.NonNull(graphene.Int, name="totalOrders")
    total_value = graphene.NonNull(graphene.Float, name="totalValue")
    average_order_value = graphene.NonNull(graphene.Float, name="averageOrderValue")


class ComplexOrdersResult(graphene.ObjectType):
    period_days = graphene.NonNull(graphene.Int, name="periodDays")
    total_users = graphene.NonNull(graphene.Int, name="totalUsers")
    data = graphene.NonNull(graphene.List(graphene.NonNull(UserOrderStats)))


class CacheEntry(graphene.ObjectType):
    key = graphene.NonNull(graphene.String)
    value = graphene.NonNull(graphene.String)
    cached = graphene.NonNull(graphene.Boolean)
    ttl = graphene.NonNull(graphene.Int)


class Query(graphene.ObjectType):
    health = graphene.NonNull(Health)
    json_items = graphene.NonNull(JsonItemsResult, limit=graphene.Int(default_value=1000), name="jsonItems")
    user = graphene.Field(User, id=graphene.NonNull(graphene.Int))
    complex_orders = graphene.NonNull(ComplexOrdersResult, days=graphene.Int(default_value=30), name="complexOrders")
    cache = graphene.NonNull(CacheEntry, key=graphene.NonNull(graphene.String))

    def resolve_health(self, info):
        db_status = check_db()
        cache_status = check_cache()
        return Health(
            status="ok",
            version="1.0.0",
            timestamp=datetime.now(timezone.utc).isoformat(),
            database=db_status,
            cache=cache_status,
        )

    def resolve_json_items(self, info, limit):
        now = datetime.now(timezone.utc).isoformat()
        items = [
            JsonItem(
                id=i,
                uuid=str(uuid.uuid4()),
                name=f"Item {i}",
                email=f"user{i}@example.com",
                created_at=now,
                is_active=i % 3 != 0,
            )
            for i in range(1, limit + 1)
        ]
        return JsonItemsResult(
            items=items,
            count=len(items),
            timestamp=now,
        )

    def resolve_user(self, info, id):
        row = get_user(id)
        if row is None:
            return None
        return User(
            id=row["id"],
            email=row["email"],
            first_name=row["firstName"],
            last_name=row["lastName"],
            age=row["age"],
            created_at=row["createdAt"],
        )

    def resolve_complex_orders(self, info, days):
        data = get_complex_orders(days)
        stats = [
            UserOrderStats(
                user_id=r["userId"],
                user_name=r["userName"],
                total_orders=r["totalOrders"],
                total_value=r["totalValue"],
                average_order_value=r["averageOrderValue"],
            )
            for r in data
        ]
        return ComplexOrdersResult(
            period_days=days,
            total_users=len(stats),
            data=stats,
        )

    def resolve_cache(self, info, key):
        value, ttl = cache_get(key)
        if value is not None:
            return CacheEntry(
                key=key,
                value=value,
                cached=True,
                ttl=ttl,
            )

        generated = f'{{"key": "{key}", "generated": true}}'
        cache_set(key, generated, 300)
        return CacheEntry(
            key=key,
            value=generated,
            cached=False,
            ttl=300,
        )


schema = graphene.Schema(query=Query)

File: src/dart/graphql/leto/lib/schema.dart
// Builds the GraphQL schema using the leto_schema API.
//
// leto_schema exposes field/object construction via the `field()` and
// `objectType()` factories (in gen.dart), which infer generics and accept a
// bare resolver function. The GraphQLObjectField constructor, by contrast,
// expects a FieldResolver wrapper, so the factories are the idiomatic entry
// point. API mappings from the classic graphql builder that this file used
// before:
//   GraphQLField(name, type:, args:, resolve:) -> field(name, type, inputs:, resolve:)
//   GraphQLFieldArg  -> GraphQLFieldInput
//   GraphQLNonNull(t)-> t.nonNull()
//   GraphQLList(t)   -> listOf(t)
//   GraphQLString/Int/Float/Boolean -> graphQLString/graphQLInt/...
//
// leto invokes a resolver as (parent, ctx): two parameters, with arguments
// reached via ctx.args rather than a third positional parameter.
import 'package:graphql_leto_benchmark/canonical.dart';
import 'package:leto_schema/leto_schema.dart';

import 'db.dart';
import 'cache.dart';

/// Leaf scalar field: resolves to parent[fieldName].
GraphQLObjectField<Object?, Object?, Object?> _leaf(
        String name, GraphQLType type) =>
    field(name, type, resolve: (obj, _) => (obj is Map) ? obj[name] : null);

GraphQLSchema buildSchema(DatabaseService db, CacheService cache) {
  final healthType = objectType('Health', fields: [
    _leaf('status', graphQLString.nonNull()),
    _leaf('version', graphQLString.nonNull()),
    _leaf('timestamp', graphQLString.nonNull()),
    _leaf('database', graphQLString.nonNull()),
    _leaf('cache', graphQLString.nonNull()),
  ]);

  final jsonItemType = objectType('JsonItem', fields: [
    _leaf('id', graphQLInt.nonNull()),
    _leaf('uuid', graphQLString.nonNull()),
    _leaf('name', graphQLString.nonNull()),
    _leaf('email', graphQLString.nonNull()),
    _leaf('createdAt', graphQLString.nonNull()),
    _leaf('isActive', graphQLBoolean.nonNull()),
  ]);

  final jsonItemsResultType = objectType('JsonItemsResult', fields: [
    _leaf('items', listOf(jsonItemType.nonNull())),
    _leaf('count', graphQLInt.nonNull()),
    _leaf('timestamp', graphQLString.nonNull()),
  ]);

  final userType = objectType('User', fields: [
    _leaf('id', graphQLInt.nonNull()),
    _leaf('email', graphQLString.nonNull()),
    _leaf('firstName', graphQLString.nonNull()),
    _leaf('lastName', graphQLString.nonNull()),
    _leaf('age', graphQLInt.nonNull()),
    _leaf('createdAt', graphQLString.nonNull()),
  ]);

  final userOrderStatsType = objectType('UserOrderStats', fields: [
    _leaf('userId', graphQLInt.nonNull()),
    _leaf('userName', graphQLString.nonNull()),
    _leaf('totalOrders', graphQLInt.nonNull()),
    _leaf('totalValue', graphQLFloat.nonNull()),
    _leaf('averageOrderValue', graphQLFloat.nonNull()),
  ]);

  final complexOrdersResultType = objectType('ComplexOrdersResult', fields: [
    _leaf('periodDays', graphQLInt.nonNull()),
    _leaf('totalUsers', graphQLInt.nonNull()),
    _leaf('data', listOf(userOrderStatsType.nonNull())),
  ]);

  final cacheEntryType = objectType('CacheEntry', fields: [
    _leaf('key', graphQLString.nonNull()),
    _leaf('value', graphQLString.nonNull()),
    _leaf('cached', graphQLBoolean.nonNull()),
    _leaf('ttl', graphQLInt.nonNull()),
  ]);

  final queryType = objectType('Query', fields: [
    field(
      'health',
      healthType.nonNull(),
      resolve: (obj, ctx) async {
        var dbStatus = 'ok';
        var cacheStatus = 'ok';
        try { await db.ping(); } catch (_) { dbStatus = 'error'; }
        try { await cache.ping(); } catch (_) { cacheStatus = 'error'; }
        return {
          'status': 'ok',
          'version': '1.0.0',
          'timestamp': DateTime.now().toUtc().toIso8601String(),
          'database': dbStatus,
          'cache': cacheStatus,
        };
      },
    ),
    field(
      'jsonItems',
      jsonItemsResultType.nonNull(),
      inputs: [GraphQLFieldInput('limit', graphQLInt, defaultValue: 1000)],
      resolve: (obj, ctx) {
        final count = itemCount(ctx.args['limit'] as int?);
        final now = DateTime.now().toUtc().toIso8601String();
        final items = List.generate(count, (i) => {
          'id': i,
          'uuid': canonicalUuid(i),
          'name': canonicalName(i),
          'email': canonicalEmail(i),
          'createdAt': canonicalCreatedAt,
          'isActive': canonicalIsActive(i),
        });
        return {
          'items': items,
          'count': items.length,
          'timestamp': now,
        };
      },
    ),
    field(
      'user',
      userType,
      inputs: [GraphQLFieldInput('id', graphQLInt.nonNull())],
      resolve: (obj, ctx) async {
        return await db.getUser(ctx.args['id'] as int);
      },
    ),
    field(
      'complexOrders',
      complexOrdersResultType.nonNull(),
      inputs: [GraphQLFieldInput('days', graphQLInt, defaultValue: 30)],
      resolve: (obj, ctx) async {
        final days = ctx.args['days'] as int? ?? 30;
        final data = await db.getComplexOrders(days);
        return {
          'periodDays': days,
          'totalUsers': data.length,
          'data': data,
        };
      },
    ),
    field(
      'cache',
      cacheEntryType.nonNull(),
      inputs: [GraphQLFieldInput('key', graphQLString.nonNull())],
      resolve: (obj, ctx) async {
        final key = ctx.args['key'] as String;
        final value = await cache.get(key);
        if (value != null) {
          final ttlVal = await cache.ttl(key);
          return {
            'key': key,
            'value': value.toString(),
            'cached': true,
            'ttl': ttlVal >= 0 ? ttlVal : 0,
          };
        }
        final generated = '{"key": "$key", "generated": true}';
        await cache.set(key, generated, 300);
        return {
          'key': key,
          'value': generated,
          'cached': false,
          'ttl': 300,
        };
      },
    ),
  ]);

  return GraphQLSchema(queryType: queryType);
}
